staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go

```diff
@@ -300,23 +310,42 @@ func toStatusErr(name string, result *metav1.Status) *apierrors.StatusError {
 }
 
 func (a *GenericAdmissionWebhook) hookClient(h *v1alpha1.ExternalAdmissionHook) (*rest.RESTClient, error) {
-	serverName := h.ClientConfig.Service.Name + "." + h.ClientConfig.Service.Namespace + ".svc"
-	u, err := a.serviceResolver.ResolveEndpoint(h.ClientConfig.Service.Namespace, h.ClientConfig.Service.Name)
+	cacheKey, err := json.Marshal(h.ClientConfig)
 	if err != nil {
 		return nil, err
 	}
+	if client, ok := a.cache.Get(string(cacheKey)); ok {
+		return client.(*rest.RESTClient), nil
+	}
 
-	// TODO: cache these instead of constructing one each time
+	serverName := h.ClientConfig.Service.Name + "." + h.ClientConfig.Service.Namespace + ".svc"
 	restConfig, err := a.authInfoResolver.ClientConfigFor(serverName)
 	if err != nil {
 		return nil, err
 	}
+
 	cfg := rest.CopyConfig(restConfig)
-	cfg.Host = u.Host
-	cfg.APIPath = path.Join(u.Path, h.ClientConfig.URLPath)
+	cfg.Host = "https://" + serverName + ":443"
```

Comment:
I'm confused. Is this field still useful? It seems the connection is always going to target at `u.Host` returned by the serviceResolver.

Reply:
I think it is no use at all.

Reply:
@liggitt What about you?

Reply:
@caesarxuchao May be to log it is useful.

Reply:
Scheme is important, it makes the connection use TLS over the dialer connection.

 I'd maybe expect the dial func to only use the service resolver when asked to dial serverName:443, and pass through for other IP/hosts

Reply:
I see.

Reply:
Why not resove for other IP/hosts?

Reply:
Here what we resolve is always service name , because originally '		u, err := a.serviceResolver.ResolveEndpoint(h.ClientConfig.Service.Namespace, h.ClientConfig.Service.Name)
' is used.

Reply:
I am confused.

Reply:
The client starts its request at the endpoint resolved by the service resolver, but any subsequent redirects or requests are handled normally. Moving the service resolution logic into a dialer (which is really where it belonged in the first place) should behave the same way:
1. Start the client pointing at the constructed service host
2. Make the custom dialer resolve the constructed service host, but pass through any dial requests for other hosts as-is

Reply:
> but any subsequent redirects or requests to other hosts are handled normally.

What's  meaning about this ?  `func (a *GenericAdmissionWebhook) hookClient(h *v1alpha1.ExternalAdmissionHook) (*rest.RESTClient, error) {
` always construct client for Service.Namespace + Name .  Services are all resolved each time start a request.  So I still do not understand  `but any subsequent redirects or requests to other hosts are handled normally.`

Reply:
Can you show me your intention with code if possible ?

Reply:
@liggitt Is this your intention? Because I see `tlsTransportCache` in k8s.io\kubernetes\staging\src\k8s.io\client-go\transport\cache.go
```
	cfg.Dial = func(network, addr string) (net.Conn, error) {
		if addr == serverName+":443" {
			u, err := a.serviceResolver.ResolveEndpoint(h.ClientConfig.Service.Namespace, h.ClientConfig.Service.Name)
			if err != nil {
				return nil, err
			}
			addr = u.Host
		}
		conn, err := net.Dial(network, addr)
		if err != nil {
			return nil, err
		}
		return conn, nil
	}
```

Reply:
@liggitt Need your comments now.

Reply:
Yes, along those lines (can build serverName+":443" outside the dialer and reuse it to compare)

Reply:
Ok, i will update and push again.

Reply:
@liggitt Done